File: DailyBrain/cpp/src/commute.cpp
// cpp/src/commute.cpp
#include "commute.h"
#include <cmath>
#include <sstream>
#include <iomanip>

double CommuteAlgorithm::haversine(const Coordinates& a, const Coordinates& b) {
    static const double R = 6371.0; // km
    auto deg2rad = [](double d) { return d * M_PI / 180.0; };

    double dlat = deg2rad(b.lat - a.lat);
    double dlon = deg2rad(b.lon - a.lon);
    double lat1 = deg2rad(a.lat);
    double lat2 = deg2rad(b.lat);

    double h = std::sin(dlat / 2) * std::sin(dlat / 2)
             + std::cos(lat1) * std::cos(lat2)
             * std::sin(dlon / 2) * std::sin(dlon / 2);

    return 2 * R * std::asin(std::sqrt(h));
}

CommuteResult CommuteAlgorithm::estimate_commute(const CommuteContext& ctx) {
    double distance_km = haversine(ctx.home, ctx.work);
    double hours = distance_km / 40.0;
    auto secs = std::chrono::seconds(static_cast<int>(hours * 3600.0));

    auto eta = ctx.departure_time + secs;

    std::time_t eta_time_t = std::chrono::system_clock::to_time_t(eta);
    std::tm tm = *std::localtime(&eta_time_t);

    std::ostringstream oss;
    oss << "Distance ~" << std::fixed << std::setprecision(1)
        << distance_km << " km, ETA at "
        << std::setfill('0') << std::setw(2) << tm.tm_hour << ":"
        << std::setfill('0') << std::setw(2) << tm.tm_min;

    CommuteResult result;
    result.estimated_duration = secs;
    result.eta = eta;
    result.note = oss.str();
    return result;
}

File: DailyBrain/rust/src/commute.rs
// rust/src/commute.rs
use std::time::Duration;
use chrono::{DateTime, Local};

#[derive(Clone, Copy, Debug)]
pub struct Coordinates {
    pub lat: f64,
    pub lon: f64,
}

pub struct CommuteContext {
    pub home: Coordinates,
    pub work: Coordinates,
    pub departure_time: DateTime<Local>,
}

pub struct CommuteResult {
    pub estimated_duration: Duration,
    pub eta: DateTime<Local>,
    pub note: String,
}

pub struct CommuteAlgorithm;

impl CommuteAlgorithm {
    pub fn estimate_commute(&self, ctx: &CommuteContext) -> CommuteResult {
        let distance_km = Self::haversine(ctx.home, ctx.work);
        let hours = distance_km / 40.0_f64;
        let secs = (hours * 3600.0_f64) as u64;
        let duration = Duration::from_secs(secs);
        let eta = ctx.departure_time + chrono::Duration::from_std(duration).unwrap();

        let note = format!("Distance ~{:.1} km, ETA at {}", distance_km, eta.format("%H:%M"));

        CommuteResult {
            estimated_duration: duration,
            eta,
            note,
        }
    }

    fn haversine(a: Coordinates, b: Coordinates) -> f64 {
        let r = 6371.0_f64; // km
        let dlat = (b.lat - a.lat).to_radians();
        let dlon = (b.lon - a.lon).to_radians();
        let lat1 = a.lat.to_radians();
        let lat2 = b.lat.to_radians();

        let h = (dlat / 2.0).sin().powi(2)
            + lat1.cos() * lat2.cos() * (dlon / 2.0).sin().powi(2);
        2.0 * r * (h.sqrt().asin())
    }
}
